artifacts/reader.py

```diff
@@ -72,6 +72,10 @@ def _ReadArtifactDefinition(self, yaml_definition):
     if not yaml_definition:
       raise errors.FormatError(u'Missing YAML definition.')
 
+    different_keys = set(yaml_definition) - definitions.TOP_LEVEL_KEYS
+    if different_keys:
+      raise errors.FormatError(u'Bad keys: {0!s}'.format(different_keys))
```

Comment:
The intention here is to catch mistyped optional keys like "urps" instead of "urls". This doesn't signal a reluctance to add additional keys in the future, just that they should be strictly defined in the spec so we can do proper validation. I think it's fine to describe any key not in the spec as "bad", but I'll change the error text here if you like. This also isn't a format change so I don't think there's any need to discuss with others. If people want to have private artifacts with their own custom keys they should override TOP_LEVEL_KEYS to include them and get proper validation.

{0:s} doesn't work for sets, you need to do the str() call first, which is what {0!s} does. Again, my preference is not to have the explicit conversion for strings, the language isn't designed to require it, and string conversion is almost always what you want. If we weren't being explicit it would just work.